### ISSUE TITLE
Encryption Issues - Security Risk

### ISSUE THREAD

**Issue by hecon5** [joyfullservice/msaccess-vcs-integration#193]
I don't (personally) use the encryption, but if you're going to, here are some notes that may help. I also really like that we can now select a hashing algorithm of choice; this is excellent!

1. The encryption algorithm should either be selected by the user as a separate option, or use the same option as the hashing algorithm. RC4 isn't really viable, and is largely only useful for obfuscation at this point.
2. The key should never be stored in plaintext *anywhere*. Users should store this separately (KeePass, 1Password, etc.). Storing keys in the registry is a fine way to ensure your data is stolen and irrecoverable once your machine dies. I recognize the convenience of this, perhaps instead of storing it in the registry, a session key could be stored momentarily and destroyed once the addin is closed.
3. If we were really advanced, we could integrate with gPGP/smartCards, etc. This is not required for this, methinks, and would cause headaches with multi-OS support.

There is an encryption routine in the `bcrypt.dll` has [encryption](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptencrypt) and [decryption ](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptdecrypt)routines; I suggest using those since we're using it for hashing. It also has [key generation (with secret) ](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgeneratesymmetrickey). 

This will ensure that encryption/hashing has a single module to manage, is handled the same, and has security patches applied via the OS level; this abstracts away a lot of the intrusion vectors that may be present.

**Comment by joyfullservice** [comment]
Good points. 👍 

The original purpose of "encryption" was to allow developers to collaborate on a project that might be stored on an external (public or private) repository, allowing both developers to build a project from source, including the encrypted content.

We have removed a lot of this need through implementing relative paths, but there is obviously still a lot of potential for sensitive information to be committed to version control. Ultimately, it needs to be up to the individual developer to _make the determination_ that what they are committing to a repository is safe to do so from a security perspective.

In the current add-in they can also incorporate a custom function in the **Run Sub After Export** to do any additional data cleaning/purging before committing to a version control system. I personally no longer use the "Encryption" feature because for my projects the security concerns were solved through the relative paths.

I could see at least two potential approaches we could take from here:

1. **Remove the "Encryption" feature entirely**, since the need for it has largely been met through the relative path conversion, and possibly custom functions in _Run Sub After Export_. This would reduce complexity and the [cost of carry](https://martinfowler.com/bliki/Yagni.html).

2. **Refactor the encryption** to use the `bcrypt.dll` functions, and remove the key storage in favor of a password prompt, or at least provide the option of not saving a password.

Since I don't currently use this feature, it's not personally a priority for me. If someone would like to tackle that project, I would be happy to review a pull request.  😄

**Comment by A9G-Data-Droid** [comment]
I have looked in to AES encryption for VBA. Although it's possible it's never going to be a good implementation unless it happens 100% outside of VBA. For this reason I would vote for option 1. It's outside the scope of this project. If you want to secure your source then use am appropriate tool for your environment. Everyone has different security frameworks with different requirements.

Sometimes a false sense of security is worse than plaintext. People might trust something they otherwise would not. Do not upload anything sensitive to a public repository. This might be easier said than done. It's an issue that plagues large dev teams. Humans make mistakes. This won't save you from yourself.

**Comment by hecon5** [comment]
I, too, am in favor of option 1: excise it from the project. External whole-package encryption is a much better (and more secure/non-reliant on ensuring it operates for all), and would ensure any other dev stuff in the folder is also encrypted. 

It's so far out of scope for an import/export tool and so dependent on the environment and security requirements therein, I think this is a case where no security is much better than some.

**Comment by joyfullservice** [comment]
Sounds great! Option one it is then. That will also be nice in the sense that it is a lot easier to carve something out than to add something new.

**Comment by hecon5** [comment]
I was thinking about this last night. We might have a sticky wicket here because the feature exists.

If we do excise it (which we should); should we include an install message that users who have encrypted it will need to decrypt before they "upgrade"? Though, if they have the already built database, they could just export that and they'd be set.

Should we have a "check for encryption key" routine on install to prevent users from installing the new version, or just alert them about this significant change?

IIRC, the `security` option states whether there is encryption in use or not; if that's set, we should nag users when they export that:
1. They are about to export in *plaintext* and that they will *not* have encryption active. 
2. If they attempt to import an encrypted system, they can't because the decryption is no longer present. 
3. The new version should let users know that the key is still stored in the registry, and ask them if they want to delete it (after decrypting their database, of course).

Alternately: we could keep the decryption portions for the next release, and only allow decryption, but not export encrypted. I don't like that option, because they could simply download the prior release and be on their way.

Thoughts?

**Comment by A9G-Data-Droid** [comment]
I agree, if the user agrees to install and export plaintext it should do a full export and wipe the old key from registry. Then it will be clean slate. You never could do source control on encrypted files. You wouldn't be able to see changes or perform merge\pull. Their old encrypted commits are of little value. In place of the decrypt function it should have a pointer to the last version that supports encryption. I wonder how many people turned this feature on.

**Comment by joyfullservice** [comment]
Thank you both for the additional thoughts and considerations. Thankfully, all of the "encryption" is in various pieces of the exported source code, so users would have no problem exporting source from their project using the updated add-in. The only issue would be with those that were trying to build the database from source using the updated add-in, since it would not decrypt those values.

Since an encryption key can be used across multiple projects, I am really hesitant to just blow it away (even after prompting the user) since there is no way to recover it if the user later finds another project that used the old key. I am not concerned about preserving any of the existing decryption functionality because the user can simply install an older version of the add-in to rebuild a database with encrypted values.

It is also a very good point to consider that probably very few people are actually using this feature. I developed it for my own purposes and used it for a time with my development team, but no longer use it. I don't know that I have heard reports of any other end users actually using it.

Here is what I would propose:

**Phase 1 - Effective Immediately**
- Remove the encryption/decryption feature from the add-in as discussed above.
- If encryption is found on source files during a build, warn user that it has been discontinued, and direct them to the wiki page for more information. The build will continue, but some objects may have errors due to invalid values.
- Add a prominent notice on the encryption wiki page explaining that this functionality has been discontinued, and that they can download (specific build) if they are needing to upgrade a project.
- Leave the existing key(s) in the registry so they will still work with prior versions of the add-in.

**Phase 2 - A Year Later**
- Remove existing encryption registry entries (with user confirmation) when installing the add-in.
- Remove wiki page on encryption. (Any users that have problems can open an issue.)
- Remove encryption check during build.

After Phase 2, we should not have any remaining traces of the encryption feature remaining, other than the verification for removal of existing keys.

**Comment by hecon5** [comment]
This is a good plan, and reduces the overhead required to remove the feature (also a consideration vice leaving as-is).

Tally ho, I say!

**Comment by hecon5** [comment]
I tweaked the Wiki Page.

**Comment by hecon5** [comment]
As an aside; when the encryption tools are excised from `dev`, I suggest this is a good time to do another release and reduce the chance of anyone using it going forward.

**Comment by joyfullservice** [comment]
Rolled this into `master` and uploaded a general release.
